### PR TITLE
[Feat][RayCluster] Use a new RayClusterReplicaFailure condition to reflect the result of reconcilePods

### PR DESCRIPTION
**Pull request by rueian**
This is a revised draft of https://github.com/ray-project/kuberay/pull/2245 based on the review.

In this version, a `RayClusterReplicaFailure` condition will be set whenever the `reconcilePods()` returns a `reconcileErr`. This is done by the `calculateStatus()` based on the `reconcileErr` solely. It assumes the `reconcilePods()` is the last reconciliation function.

We also consolidate the fine-grained condition reasons in the previous draft into one `FailedReconcilePods` for simplicity.

Tests are conducted in three parts:
1. Assert the `reconcilePods()` will return a `reconcileErr`.
2. Assert the `calculateStatus()` will set the condition only when the feature gate is enabled.
3. Assert the `inconsistentRayClusterStatus()` will report `true` when conditions are changed.

## Checks

- [x] I've made sure the tests are passing.
- Testing Strategy
   - [x] Unit tests
   - [x] Manual tests

### CODE REVIEW

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 1]
```diff
@@ -1149,6 +1168,22 @@ func (r *RayClusterReconciler) calculateStatus(ctx context.Context, instance *ra
 	// Deep copy the instance, so we don't mutate the original object.
 	newInstance := instance.DeepCopy()
 
+	if features.Enabled(features.RayClusterStatusConditions) {
+		if reconcileErr != nil {
+			if errstd.Is(reconcileErr, reconcilePodsErr) {
+				meta.SetStatusCondition(&newInstance.Status.Conditions, metav1.Condition{
+					Type:    string(rayv1.RayClusterReplicaFailure),
+					Status:  metav1.ConditionTrue,
+					Reason:  "FailedReconcilePods",
```
Comment (andrewsykim) [inline ray-operator/controllers/ray/raycluster_controller.go:1177]: Failing to reconcile pods is different from replica failure, do we need to make this distinction clearer? 

For example, replica failure could mean one of the pods are unhealthy and need to be recreated, but a reconcile error could happen for other reasons before the pod is running

Comment (rueian) [reply ray-operator/controllers/ray/raycluster_controller.go:1177]: Hi @andrewsykim,

Possible distinctions are those in the previous [draft](https://github.com/ray-project/kuberay/pull/2245):
* FailedDeleteAllPods
* FailedDeleteHeadPod
* FailedCreateHeadPod
* FailedDeleteWorkerPod
* FailedCreateWorkerPod

Do they make sense to you?

Comment (andrewsykim) [reply ray-operator/controllers/ray/raycluster_controller.go:1177]: Yes, but my point is that maybe reconcile error and replica failures are better off as separate conditions instead of different reasons within the same condition.

Comment (rueian) [reply ray-operator/controllers/ray/raycluster_controller.go:1177]: We borrow the idea of the replica failures condition from the Kubernetes ReplicaSet controller. It has only two possible reasons:
* FailedCreate
* FailedDelete

And it doesn't reflect the health of pods either. I guess that was because the health of pods is handled by the pod controller already.

The `FailedReconcilePods` reason in this PR may be oversimplified and exceeds the semantic of `ReplicaFailure` indeed. I think I can bring the previous draft back by introducing them as individual error instances and wrapping them at the appropriate points. However, those points are still in the `reconcilePods` function.

``` go
var (
	failedDeleteAllPods   = errstd.New("...")
	failedDeleteHeadPod   = errstd.New("...")
	failedCreateHeadPod   = errstd.New("...")
	failedDeleteWorkerPod = errstd.New("...")
	failedCreateWorkerPod = errstd.New("...")
)

func (r *RayClusterReconciler) reconcilePods(ctx context.Context, instance *rayv1.RayCluster) error {
		...
		if err := r.createHeadPod(ctx, *instance); err != nil {
			common.FailedClustersCounterInc(instance.Namespace)
			return errstd.Join(failedCreateHeadPod, err)
		}
		...
}
```

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 2]
```diff
@@ -591,7 +600,17 @@ func (r *RayClusterReconciler) reconcileHeadlessService(ctx context.Context, ins
 	return nil
 }
 
+// reconcilePodsErr is a marker used by the calculateStatus() for setting the RayClusterReplicaFailure condition.
+var reconcilePodsErr = errstd.New("reconcile pods error")
```
Comment (kevin85421) [inline ray-operator/controllers/ray/raycluster_controller.go:604]: move to `constant.go`?

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 3]
```diff
@@ -391,6 +394,12 @@ func (r *RayClusterReconciler) inconsistentRayClusterStatus(ctx context.Context,
 			oldStatus.Endpoints, newStatus.Endpoints, oldStatus.Head, newStatus.Head))
 		return true
 	}
+	if !reflect.DeepEqual(oldStatus.Conditions, newStatus.Conditions) {
+		logger.Info("inconsistentRayClusterStatus", "detect inconsistency", fmt.Sprintf(
```
Comment (kevin85421) [inline ray-operator/controllers/ray/raycluster_controller.go:398]: Avoid using `fmt.Sprintf`. Please use 

```go
logger.Info(..., ..., "old conditions", oldStatus.Conditions, "new conditions", newStatus.Conditions)
```

instead. See [this doc](https://github.com/nginxinc/nginx-gateway-fabric/blob/main/docs/developer/logging-guidelines.md#message-guidelines) for more details.

File: ray-operator/controllers/ray/raycluster_controller.go (hunk 1 above)
Comment (kevin85421) [reply ray-operator/controllers/ray/raycluster_controller.go:1183]:
```go
var (
	failedDeleteAllPods   = errstd.New("...")
	failedDeleteHeadPod   = errstd.New("...")
	failedCreateHeadPod   = errstd.New("...")
	failedDeleteWorkerPod = errstd.New("...")
	failedCreateWorkerPod = errstd.New("...")
)
...
```

This makes sense to me.

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 4]
```diff
@@ -591,7 +600,17 @@ func (r *RayClusterReconciler) reconcileHeadlessService(ctx context.Context, ins
 	return nil
 }
 
+// reconcilePodsErr is a marker used by the calculateStatus() for setting the RayClusterReplicaFailure condition.
+var reconcilePodsErr = errstd.New("reconcile pods error")
+
 func (r *RayClusterReconciler) reconcilePods(ctx context.Context, instance *rayv1.RayCluster) error {
+	if err := r.doReconcilePods(ctx, instance); err != nil {
+		return fmt.Errorf("%w: %w", reconcilePodsErr, err)
```
[the remark this answers is not in the document]
Comment (rueian) [reply ray-operator/controllers/ray/raycluster_controller.go]: We can use Join. But Unwrap doesn't work well with the Join.
image

[the remark this answers is not in the document]
Comment (rueian) [reply ray-operator/controllers/ray/raycluster_controller.go]: To unwrap a joined error, we need a special trick:
```go
errs := err.(interface{ Unwrap() []error }).Unwrap()
```

File: ray-operator/controllers/ray/utils/util_test.go [hunk 5]
```diff
@@ -523,3 +523,11 @@ env_vars:
 		})
 	}
 }
+
+func TestErrRayClusterReplicaFailureReason(t *testing.T) {
```
Comment (kevin85421) [inline ray-operator/controllers/ray/utils/util_test.go:527]: can we test the case that the error is not a `RayClusterReplicaFailure`?

**Review by kevin85421 — APPROVED** [review]
LGTM. Just left a nit comment.

File: ray-operator/controllers/ray/utils/util_test.go (hunk 5 above)
Comment (rueian) [reply ray-operator/controllers/ray/utils/util_test.go:528]: Sure. I will add a test that the RayClusterReplicaFailureReason() should return an empty string for a random error.
